### PR TITLE
Fix: `nn.Parameter` return type identified as `Tensor` instead of `nn.Parameter`

### LINKED ISSUE #125105: [typing] `nn.Parameter` return type identified as `Tensor` by `pyright`
### 🐛 Describe the bug

Type checking the following script with `pyright` fails (but not with `mypy`):

```python
from typing import assert_type, reveal_type
import torch

w = torch.randn(3)
p = torch.nn.Parameter(w)
assert_type(p, torch.nn.Parameter)  # ❌ expected "Parameter" but received "Tensor"
reveal_type(torch.nn.Parameter.__new__)
# (self: type[Self@TensorBase], *args: Unknown, **kwargs: Unknown) -> Tensor
```

My guess what's happening:

1. `TensorBase` defines: `def __new__(self, *args, **kwargs) -> Tensor: ...` in the stub `torch/_C/__init__.pyi`
2. `Parameter.__new__` is untyped. Both `pyright` and `mypy` show it's `self: type[Self@TensorBase] -> Tensor`, so reusing the superclasses type.
3. `mypy` ignores the discrepancy and treats the result as an `nn.Paramter` anyway.

It seems the issue can be fixed by simply changing the signature in `TensorBase.__new__` to 

```python
def __new__(cls, *args, **kwargs) -> Self
```

### Versions

- torch 2.3.0
- pyright 1.1.360

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #125105

cc @ezyang @malfet @xuzhao9 @gramster @albanD

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by randolf-scholz** [comment]
@ezyang There is no release note label for type hint changes, which one should I use? nn?

File: tools/pyi/gen_pyi.py [hunk 1]
```diff
@@ -1064,14 +1064,14 @@ def replace_special_case(hint: str) -> str:
             "new_tensor": [
                 f"def new_tensor(self, data: Any, {FACTORY_PARAMS}) -> Tensor: ..."
             ],
-            "__new__": ["def __new__(self, *args, **kwargs) -> Tensor: ..."],
+            "__new__": ["def __new__(cls, *args, **kwargs) -> Self: ..."],
```
[the remark this answers is not in the document]
Comment (randolf-scholz) [reply tools/pyi/gen_pyi.py:1067]: Hm, I tried with your example https://github.com/albanD/subclass_zoo/blob/main/uint4_tensor.py, and `Tensor.__new__(UInt4Tensor)` gives `UInt4Tensor([])` of type `<class 'uint4_tensor.UInt4Tensor'>`

[the remark this answers is not in the document]
Comment (randolf-scholz) [reply tools/pyi/gen_pyi.py:1067]:
```python
from torch import Tensor

class MyTensor(torch.Tensor): ...

t = Tensor.__new__(MyTensor, [1,2,3])
assert type(t) == MyTensor  # ✅
```

I don't see the issue...

[the remark this answers is not in the document]
Comment (ezyang) [reply tools/pyi/gen_pyi.py:1067]: I think it's reasonable to make sure we have accurate types for Tensor and Parameter only though, surely?!

[the remark this answers is not in the document]
Comment (albanD) [reply tools/pyi/gen_pyi.py:1067]: What I meant is this:
```python
from torch.nn import Parameter
from torch import Tensor

class MyTensor(Tensor): ...

t = Parameter(MyTensor([1,2,3]))
assert type(t) == MyTensor
assert type(t) != Parameter
```

[the remark this answers is not in the document]
Comment (randolf-scholz) [reply tools/pyi/gen_pyi.py:1067]: Right, here's what the patch changes, when using `mypy`/`pyright`:

```python
from torch.nn import Parameter
from torch import Tensor

class MyTensor(Tensor): ...

t = Parameter(MyTensor([1,2,3]))
reveal_type(t)  # before/after patch: mypy: Parameter/Parameter; pyright: Tensor/Parameter
assert type(t) == MyTensor
assert type(t) != Parameter

t2 = Parameter(Tensor([1,2,3]))
reveal_type(t2)  # before/after patch: mypy: Parameter/Parameter, pyright: Tensor/Parameter
assert type(t2) != Tensor
assert type(t2) == Parameter
```

The patch does not solve all problems, but it aligns the typing of `mypy` and `pyright` in the most important case. There is no behavioral change when using `mypy`.

I think solving the issue for `nn.Parameter` for custom Tensors is difficult with the current python type system, because the behavior violates the Liskov substitution principle and the type system does not cater to such cases.

[the remark this answers is not in the document]
Comment (randolf-scholz) [reply tools/pyi/gen_pyi.py:1067]: And again, to clarify: This is because `mypy` actually ignores the annotated return type of `__new__`, but `pyright` doesn't.

```python
class A:
    def __new__(cls) -> int: return 0
        
reveal_type(A())  # mypy: A, pyright: int
```

- [mypy-playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=e975c6d4cc64e6390f2fe6e6e067fd2f)
- [pyright-playground](https://pyright-play.net/?code=MYGwhgzhAECCBcAoaLoBMCmAzaB9XAdhgO74AUoEAlNALQB80AlgQC7zQBOGrArpwWgAGZKjGJuANwxgQuVgE8ADhjKwyVGtADE0ALbKFHWABpoShZyYBzABbtmbIA)
- https://github.com/python/mypy/issues/1020

**Review by albanD — APPROVED** [review]
This doesn't effect runtime, so I'm happy with doing it even if it is only a partial fix!
